File: src/gitDiff.ts
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { loadSpec, loadSpecFromContent, SpecLoadError } from "./loadSpec.js";
import type { PostulateSpec } from "./spec.js";

export class GitDiffError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitDiffError";
  }
}

function isEnoent(err: unknown): boolean {
  return (
    typeof err === "object" &&
    err !== null &&
    "code" in err &&
    (err as { code?: string }).code === "ENOENT"
  );
}

function runGit(
  args: string[],
  cwd: string
): { ok: boolean; stdout: string; stderr: string } {
  const result = spawnSync("git", args, { cwd, encoding: "utf8" });
  if (result.error) {
    if (isEnoent(result.error)) {
      throw new GitDiffError("Git executable not found on PATH");
    }
    throw new GitDiffError(`Could not run git: ${result.error.message}`);
  }
  return {
    ok: result.status === 0,
    stdout: result.stdout ?? "",
    stderr: result.stderr ?? ""
  };
}

export function findGitRoot(cwd: string = process.cwd()): string {
  let start: string;
  try {
    start = path.resolve(cwd);
    if (!fs.statSync(start).isDirectory()) {
      throw new GitDiffError(`Working directory is not accessible: ${start}`);
    }
  } catch (err) {
    if (err instanceof GitDiffError) throw err;
    throw new GitDiffError(
      `Working directory is not accessible: ${path.resolve(cwd)}`
    );
  }

  const result = runGit(["rev-parse", "--show-toplevel"], start);
  if (!result.ok) {
    throw new GitDiffError(
      "Not a git repository (git rev-parse --show-toplevel failed)"
    );
  }
  return result.stdout.trim();
}

function resolveSpecPath(specPath: string, cwd: string): string {
  return path.isAbsolute(specPath) ? specPath : path.resolve(cwd, specPath);
}

/** Reject paths that escape root using path components (allows `..spec.yaml`). */
function isOutsideRepo(root: string, target: string): boolean {
  const relativePath = path.relative(root, target);
  return (
    relativePath === ".." ||
    relativePath.startsWith(`..${path.sep}`) ||
    path.isAbsolute(relativePath)
  );
}

export function validateWorkingSpecFile(
  specPath: string,
  gitRoot: string
): { resolved: string; repoRelative: string } {
  let lstat: fs.Stats;
  try {
    lstat = fs.lstatSync(specPath);
  } catch (err) {
    if (isEnoent(err)) {
      throw new SpecLoadError(`Spec file not found: ${specPath}`);
    }
    throw new GitDiffError(`Could not read spec path: ${specPath}`);
  }

  if (lstat.isSymbolicLink()) {
    throw new GitDiffError(
      `Spec file must be a regular file, not a symlink: ${specPath}`
    );
  }
  if (!lstat.isFile()) {
    throw new GitDiffError(`Spec path is not a regular file: ${specPath}`);
  }

  const resolved = fs.realpathSync(specPath);
  const root = fs.realpathSync(gitRoot);
  if (isOutsideRepo(root, resolved)) {
    throw new GitDiffError(
      `Spec file ${specPath} is outside the git repository at ${gitRoot}`
    );
  }

  const repoRelative = path.relative(root, resolved).split(path.sep).join("/");
  return { resolved, repoRelative };
}

export function resolveCommit(ref: string, cwd: string): string {
  const trimmed = ref.trim();
  if (!trimmed) {
    throw new GitDiffError("Git ref must not be empty");
  }

  const result = runGit(
    ["rev-parse", "--verify", "--end-of-options", `${trimmed}^{commit}`],
    cwd
  );
  if (!result.ok) {
    throw new GitDiffError(`Git ref not found: ${ref}`);
  }
  return result.stdout.trim();
}

function gitShowSpecAtCommit(
  commit: string,
  repoRelativePath: string,
  cwd: string
): string {
  const exists = runGit(["cat-file", "-e", `${commit}:${repoRelativePath}`], cwd);
  if (!exists.ok) {
    throw new GitDiffError(`Spec not found at ${commit}:${repoRelativePath}`);
  }

  const result = runGit(["show", `${commit}:${repoRelativePath}`], cwd);
  if (!result.ok) {
    throw new GitDiffError(
      `Could not read spec at ${commit}:${repoRelativePath}`
    );
  }
  return result.stdout;
}

export function loadSpecAtGitRef(
  ref: string,
  specPath: string,
  cwd: string = process.cwd()
): PostulateSpec {
  const invocationCwd = path.resolve(cwd);
  const gitRoot = findGitRoot(invocationCwd);
  const resolvedSpec = resolveSpecPath(specPath, invocationCwd);
  const { repoRelative } = validateWorkingSpecFile(resolvedSpec, gitRoot);
  const commit = resolveCommit(ref, gitRoot);
  const source = `${commit}:${repoRelative}`;
  try {
    const content = gitShowSpecAtCommit(commit, repoRelative, gitRoot);
    return loadSpecFromContent(content, source);
  } catch (err) {
    if (err instanceof GitDiffError) {
      throw new SpecLoadError(err.message);
    }
    throw err;
  }
}

export function loadSpecsForGitDiff(
  ref: string,
  specPath: string,
  cwd: string = process.cwd()
): { before: PostulateSpec; after: PostulateSpec } {
  const invocationCwd = path.resolve(cwd);
  const gitRoot = findGitRoot(invocationCwd);
  const resolvedSpec = resolveSpecPath(specPath, invocationCwd);
  const { resolved, repoRelative } = validateWorkingSpecFile(
    resolvedSpec,
    gitRoot
  );
  const commit = resolveCommit(ref, gitRoot);
  let content: string;
  try {
    content = gitShowSpecAtCommit(commit, repoRelative, gitRoot);
  } catch (err) {
    if (err instanceof GitDiffError) {
      throw new SpecLoadError(err.message);
    }
    throw err;
  }
  const before = loadSpecFromContent(content, `${commit}:${repoRelative}`);
  const after = loadSpec(resolved);
  return { before, after };
}

File: adapters/python/src/postulate/git_diff.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path

from postulate.load_spec import SpecLoadError, load_spec, load_spec_from_content
from postulate.models import PostulateSpec


class GitDiffError(Exception):
    """Raised when git-aware diff prerequisites fail."""


def _run_git(args: list[str], cwd: Path) -> subprocess.CompletedProcess[str]:
    try:
        return subprocess.run(
            ["git", *args],
            cwd=cwd,
            capture_output=True,
            text=True,
            check=False,
        )
    except FileNotFoundError as err:
        raise GitDiffError("Git executable not found on PATH") from err
    except OSError as err:
        raise GitDiffError(f"Could not run git: {err}") from err


def find_git_root(cwd: Path | None = None) -> Path:
    start = (cwd or Path.cwd()).resolve()
    if not start.is_dir():
        raise GitDiffError(f"Working directory is not accessible: {start}")

    result = _run_git(["rev-parse", "--show-toplevel"], start)
    if result.returncode != 0:
        raise GitDiffError(
            "Not a git repository (git rev-parse --show-toplevel failed)"
        )
    return Path(result.stdout.strip())


def resolve_spec_path(spec_path: Path | str, cwd: Path) -> Path:
    path = Path(spec_path)
    if not path.is_absolute():
        path = cwd / path
    return path


def validate_working_spec_file(
    spec_path: Path,
    git_root: Path,
) -> tuple[Path, str]:
    if os.path.lexists(spec_path) and spec_path.is_symlink():
        raise GitDiffError(
            f"Spec file must be a regular file, not a symlink: {spec_path}"
        )

    if not spec_path.exists():
        raise SpecLoadError(f"Spec file not found: {spec_path}")

    if not spec_path.is_file():
        raise GitDiffError(f"Spec path is not a regular file: {spec_path}")

    resolved = Path(os.path.realpath(spec_path))
    root = Path(os.path.realpath(git_root))
    try:
        repo_relative = resolved.relative_to(root).as_posix()
    except ValueError as err:
        raise GitDiffError(
            f"Spec file {spec_path} is outside the git repository at {git_root}"
        ) from err

    return resolved, repo_relative


def resolve_commit(ref: str, cwd: Path) -> str:
    trimmed = ref.strip()
    if not trimmed:
        raise GitDiffError("Git ref must not be empty")

    result = _run_git(
        ["rev-parse", "--verify", "--end-of-options", f"{trimmed}^{{commit}}"],
        cwd,
    )
    if result.returncode != 0:
        raise GitDiffError(f"Git ref not found: {ref}")
    return result.stdout.strip()


def git_show_spec_at_commit(
    commit: str,
    repo_relative_path: str,
    cwd: Path,
) -> str:
    exists = _run_git(["cat-file", "-e", f"{commit}:{repo_relative_path}"], cwd)
    if exists.returncode != 0:
        raise GitDiffError(f"Spec not found at {commit}:{repo_relative_path}")

    result = _run_git(["show", f"{commit}:{repo_relative_path}"], cwd)
    if result.returncode != 0:
        raise GitDiffError(
            f"Could not read spec at {commit}:{repo_relative_path}"
        )
    return result.stdout


def repo_relative_spec_path(spec_path: Path, git_root: Path) -> str:
    _, repo_relative = validate_working_spec_file(spec_path, git_root)
    return repo_relative


def git_show_spec_at_ref(ref: str, repo_relative_path: str, cwd: Path) -> str:
    commit = resolve_commit(ref, cwd)
    return git_show_spec_at_commit(commit, repo_relative_path, cwd)


def load_spec_at_git_ref(
    ref: str,
    spec_path: Path | str,
    *,
    cwd: Path | None = None,
) -> PostulateSpec:
    invocation_cwd = (cwd or Path.cwd()).resolve()
    git_root = find_git_root(invocation_cwd)
    resolved_spec = resolve_spec_path(spec_path, invocation_cwd)
    _, repo_path = validate_working_spec_file(resolved_spec, git_root)
    commit = resolve_commit(ref, git_root)
    source = f"{commit}:{repo_path}"
    try:
        content = git_show_spec_at_commit(commit, repo_path, git_root)
    except GitDiffError as err:
        raise SpecLoadError(str(err)) from err
    return load_spec_from_content(content, source)


def diff_specs_against_git_ref(
    ref: str,
    spec_path: Path | str,
    *,
    cwd: Path | None = None,
) -> tuple[PostulateSpec, PostulateSpec]:
    invocation_cwd = (cwd or Path.cwd()).resolve()
    git_root = find_git_root(invocation_cwd)
    resolved_spec = resolve_spec_path(spec_path, invocation_cwd)
    _, repo_path = validate_working_spec_file(resolved_spec, git_root)
    commit = resolve_commit(ref, git_root)
    try:
        content = git_show_spec_at_commit(commit, repo_path, git_root)
    except GitDiffError as err:
        raise SpecLoadError(str(err)) from err
    before = load_spec_from_content(content, f"{commit}:{repo_path}")
    after = load_spec(resolved_spec)
    return before, after
